src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs

```diff
@@ -10,4 +10,402 @@
+
+moveToFront ∷ (Env.Instruction m, Integral a) ⇒ a → m Instr.ExpandedOp
+moveToFront num = do
+  let inst = Instructions.dig (fromIntegral num)
+  addInstr inst
+  modify @"stack" (VStack.dig (fromIntegral num))
+  pure inst
+
+dupToFront ∷ (Env.Instruction m, Integral a) ⇒ a → m Instr.ExpandedOp
+dupToFront num = do
+  modify @"stack" (VStack.dupDig (fromIntegral num))
+  let instrs =
+        [ Instructions.dig (fromIntegral num),
+          Instructions.dup,
+          Instructions.dug (fromIntegral num)
+        ]
+  addInstrs instrs
+  pure (fold instrs)
+
+data Protect
+  = Protect
+      { val ∷ Env.Expanded,
+        insts ∷ [Types.Op]
+      }
+
+protect ∷ Env.Ops m ⇒ m Env.Expanded → m Protect
+protect inst = do
+  curr ← get @"ops"
+  v ← inst
+  after ← get @"ops"
+  put @"ops" curr
+  pure Protect {val = v, insts = after}
+
+-- Promoting types happen elsewhere
+-- so protect just serves to hold the ops effects
+addExpanded ∷ Env.Ops m ⇒ Protect → m ()
+addExpanded (Protect _ i) = addInstrs i
+
+promoteTopStack ∷ Env.Instruction m ⇒ Env.Expanded → m Env.Expanded
+promoteTopStack x = do
+  stack ← get @"stack"
+  let (insts, stack') = VStack.promote 1 stack
+  put @"stack" stack'
+  addInstrs insts
+  pure x
+
+reserveNames ∷ HasState "count" Word m ⇒ Word → m [Symbol]
+reserveNames i = do
+  c ← get @"count"
+  put @"count" (i + c)
+  pure (intern . show <$> [c .. c + i - 1])
+
+-- TODO ∷ drop extra things from the vstack, mainly
+-- 1. the function we move to the front
+-- 2. the expanded lambda we create
+-- Note that it isn't vital, as these aren't stored in the real Michelson stack
+-- Other things considered:
+-- We don't need to drop the arguments we eval and name, as they should be eaten
+-- by the functions they call with the appropriate usages
+apply ∷ Env.Reduction m ⇒ Env.Curr → [Types.NewTerm] → [Symbol] → m Env.Expanded
+apply closure args remainingArgs = do
+  let total_length = fromIntegral (length args + length remainingArgs)
+  case total_length `compare` Env.left closure of
+    EQ → do
+      -- usage and type doesn't matter here!
+      evalArgsAndName
+      app
+    LT → do
+      evalArgsAndName
+      -- make new closure, and apply args to the fun in the closure
+      -- allowing more arguments to come, in a cont style!
+      let remaining = Env.left closure - total_length
+
+          (captured, left) = splitAt (fromIntegral total_length) (Env.argsLeft closure)
+
+          con = Env.Curr Env.C
+            { Env.left = remaining,
+              Env.argsLeft = left,
+              Env.captures = foldr Set.insert (Env.captures closure) captured,
+              Env.ty = Env.ty closure,
+              Env.argsApplied = [],
+              Env.fun = Env.Fun $ \args →
+                Env.unFun
+                  (Env.fun closure)
+                  (args <> fmap makeVar (reverse captured))
+            }
+
+      consVal con (Env.ty closure)
```

Comment:
Is it the type of the closure? Haven't we already applied some arguments?

Reply:
Yeah the type should be updated. Mind adding a pr to change it?

Reply:
I have since updated the type, So we dispatch to the follwoing


```haskell
          con = Env.C
            { ...
              Env.ty = eatType (fromIntegral totalLength) (Env.ty closure),
            }
      consVal (Env.Curr con) (Env.ty con)
      pure (Env.Curr con)

eatType ∷ Natural → Types.Type → Types.Type
eatType 0 t = t
eatType x (Ann.Pi _ _ a) = eatType (pred x) a
eatType _ _ = error "Only eat parts of a Pi types, not any other type!"
```
Does this look okay to you?

Reply:
Yes, that looks correct.